### PR TITLE
Support arbitrary padding values in copy constraint

### PR DESCRIPTION
**Pull request by liliatangxy**
Previously, copy constraint could only support pad values of 0. However, operations such as max and min may produce incorrect outputs when they perform operations containing both actual values of the tensor and pad values of 0.

Also fixes a bug introduced in the LSTM PR with a cq block in Conv layer

### CODE REVIEW

File: src/basic_block/copy_constraint.rs [hunk 1]
```diff
@@ -63,7 +63,7 @@ fn construct_ssig(
   idxs: &[((usize, usize), Option<(usize, usize)>)],
   N: usize,
   last_dim: usize,
-  partitions: &HashMap<Option<(usize, usize)>, Vec<(usize, usize)>>,
+  partitions: &HashMap<Option<(usize, usize)>, Vec<Vec<(usize, usize)>>>,
```
Comment (ddkang) [inline src/basic_block/copy_constraint.rs:68]: This needs some comments

**Review by ddkang — COMMENTED** [review]
I think this needs some tests

File: src/basic_block/copy_constraint.rs [hunk 2]
```diff
@@ -90,23 +89,52 @@ fn construct_ssig(
     .collect()
 }
 
+// Returns the padding_partitions field for CopyConstraintBasicBlock when the given permutation padding elements are 0
+pub fn zero_padding_partition(permutation: &ArrayD<Option<IxDyn>>) -> BTreeMap<Fr, Vec<IxDyn>> {
```
Comment (Vimalakirti) [inline src/basic_block/copy_constraint.rs:92]: Is there any reason that we don't put this in `util/`?

File: src/layer/conv.rs [hunk 3]
```diff
@@ -147,7 +152,7 @@ impl Layer for ConvLayer {
     }));
     let change_SF_check = graph.addBB(Box::new(RepeaterBasicBlock {
       basic_block: Box::new(CQBasicBlock {
-        setup: Array1::from_iter(-(1 << 10)..1 << 11).map(|x| Fr::from(*x)),
+        setup: Array1::from_iter(onnx::CQ_RANGE_LOWER..-onnx::CQ_RANGE_LOWER).map(|x| Fr::from(*x)),
```
Comment (Vimalakirti) [inline src/layer/conv.rs:155]: Is there any reason why we want this array to be in descending order?

Comment (liliatangxy) [reply src/layer/conv.rs:155]: That should be ascending order since `CQ_RANGE_LOWER` is the min

File: src/basic_block/copy_constraint.rs (hunk 2 above)
Comment (liliatangxy) [reply src/basic_block/copy_constraint.rs:92]: Wasn't sure whether it makes more sense to put it in util or here given that it is only ever used with CCBB. I can move it to util if that is more appropriate

**Comment by liliatangxy** [comment]
> I think this needs some tests

Currently there is a test that uses the `zero_padding_partition` function which creates one padding value, another test for a permutation with two different padding values, and tests with no padding values. Not sure what other tests to add

**Comment by Vimalakirti** [comment]
When we have no need to use index, we might skip `.enumerate()` after `iter`. Lgtm minus this.

**Review by Vimalakirti — APPROVED** [review]

**Review by Vimalakirti — APPROVED** [review]

**Review by ddkang — APPROVED** [review]

**Pull request merged**